File: cxx/src/indexer/mft_parser.cpp
#include "indexer/mft_parser.h"

#include <algorithm>
#include <cstring>

namespace swiftlist::indexer {

namespace {

constexpr int64_t kFileTimeToUnixOffset = 116444736000000000LL; // 1601-01-01 to 1970-01-01 in 100ns

uint64_t ReadLEImpl(const uint8_t* p, int n) {
    uint64_t v = 0;
    for (int i = 0; i < n && i < 8; ++i) {
        v |= static_cast<uint64_t>(p[i]) << (8 * i);
    }
    return v;
}

int64_t ReadSignedLEImpl(const uint8_t* p, int n) {
    int64_t v = 0;
    for (int i = 0; i < n && i < 8; ++i) {
        v |= static_cast<int64_t>(p[i]) << (8 * i);
    }
    // Sign-extend.
    if (n < 8 && (p[n - 1] & 0x80)) {
        v |= (~static_cast<int64_t>(0)) << (8 * n);
    }
    return v;
}

} // namespace

uint64_t ReadLE(const uint8_t* p, int n) {
    return ReadLEImpl(p, n);
}

int64_t ReadSignedLE(const uint8_t* p, int n) {
    return ReadSignedLEImpl(p, n);
}

uint32_t FileTimeToUnixSeconds(int64_t fileTime) {
    if (fileTime <= 0) return 0;
    int64_t unixTime = (fileTime - kFileTimeToUnixOffset) / 10'000'000LL;
    if (unixTime <= 0) return 0;
    if (unixTime > static_cast<int64_t>(UINT32_MAX)) return UINT32_MAX;
    return static_cast<uint32_t>(unixTime);
}

bool ApplyFixup(std::span<uint8_t> buf, uint32_t bytesPerSector) {
    if (bytesPerSector < 2 || buf.size() < 8) return false;

    uint16_t usaOff = static_cast<uint16_t>(ReadLE(buf.data() + 4, 2));
    uint16_t usaCount = static_cast<uint16_t>(ReadLE(buf.data() + 6, 2));

    // usaCount includes the placeholder USN at index 0.
    if (usaOff == 0 || usaCount < 2) return true; // nothing to fix

    for (uint16_t i = 1; i < usaCount; ++i) {
        size_t secEnd = static_cast<size_t>(i) * bytesPerSector - 2;
        size_t usaEntry = usaOff + i * 2;

        if (secEnd + 1 >= buf.size() || usaEntry + 1 >= buf.size()) {
            break; // out of bounds, stop
        }

        buf[secEnd]     = buf[usaEntry];
        buf[secEnd + 1] = buf[usaEntry + 1];
    }

    return true;
}

std::vector<DataRun> ParseDataRuns(std::span<const uint8_t> attributeValue) {
    std::vector<DataRun> runs;
    if (attributeValue.empty()) return runs;

    int64_t lcn = 0;
    size_t p = 0;

    while (p < attributeValue.size()) {
        uint8_t hdr = attributeValue[p++];
        if (hdr == 0) break; // end marker

        int lenBytes = hdr & 0x0F;
        int offBytes = (hdr >> 4) & 0x0F;

        if (p + lenBytes + offBytes > attributeValue.size()) break;

        uint64_t runLen = ReadLE(attributeValue.data() + p, lenBytes);
        p += lenBytes;

        int64_t runOff = ReadSignedLE(attributeValue.data() + p, offBytes);
        p += offBytes;

        lcn += runOff;
        if (runLen > 0) {
            runs.push_back({lcn, runLen});
        }
    }

    return runs;
}

std::vector<MftNameEntry> CollectNames(std::span<const uint8_t> record,
                                       uint32_t bytesPerSector,
                                       int recordIndex) {
    std::vector<MftNameEntry> names;
    if (record.size() < 24) return names;

    // Check magic.
    if (record[0] != 'F' || record[1] != 'I' || record[2] != 'L' || record[3] != 'E') {
        return names;
    }

    uint16_t flags = static_cast<uint16_t>(ReadLE(record.data() + 22, 2));
    if ((flags & 0x01) == 0) return names; // not in use

    // First attribute offset.
    uint16_t firstAttrOff = static_cast<uint16_t>(ReadLE(record.data() + 20, 2));
    size_t attrOff = firstAttrOff;
    size_t recordLen = record.size();

    // Per-record state.
    uint32_t creationUnix = 0;
    uint32_t lastWriteUnix = 0;
    uint32_t lastAccessUnix = 0;
    uint32_t fileAttrs = 0;
    uint64_t dataRealSize = UINT64_MAX; // from $DATA

    while (attrOff + 8 <= recordLen) {
        uint32_t attrType = static_cast<uint32_t>(ReadLE(record.data() + attrOff, 4));
        uint32_t attrLen = static_cast<uint32_t>(ReadLE(record.data() + attrOff + 4, 4));

        if (attrType == 0xFFFFFFFF) break; // end marker
        if (attrLen == 0 || attrOff + attrLen > recordLen) break;

        uint8_t residentFlag = record[attrOff + 8];
        uint8_t nameLen = record[attrOff + 9];
        uint16_t nameOff = static_cast<uint16_t>(ReadLE(record.data() + attrOff + 10, 2));

        if (attrType == 0x10 && residentFlag == 0) {
            // $STANDARD_INFORMATION (resident)
            uint16_t valueOff = static_cast<uint16_t>(ReadLE(record.data() + attrOff + 20, 2));
            size_t vo = attrOff + valueOff;
            if (vo + 44 <= recordLen) {
                int64_t creation = static_cast<int64_t>(ReadLE(record.data() + vo, 8));
                int64_t lastWrite = static_cast<int64_t>(ReadLE(record.data() + vo + 8, 8));
                int64_t lastAccess = static_cast<int64_t>(ReadLE(record.data() + vo + 0x18, 8));
                creationUnix = FileTimeToUnixSeconds(creation);
                lastWriteUnix = FileTimeToUnixSeconds(lastWrite);
                lastAccessUnix = FileTimeToUnixSeconds(lastAccess);
                fileAttrs = static_cast<uint32_t>(ReadLE(record.data() + vo + 0x20, 4));
            }
        } else if (attrType == 0x30 && residentFlag == 0) {
            // $FILE_NAME (resident)
            uint16_t valueOff = static_cast<uint16_t>(ReadLE(record.data() + attrOff + 20, 2));
            size_t vo = attrOff + valueOff;
            if (vo + 66 <= recordLen) {
                UInt128 parentFrn{ReadLE(record.data() + vo, 8), 0};
                uint64_t realSize = ReadLE(record.data() + vo + 0x30, 8);
                uint8_t nameLength = record[vo + 0x40];
                uint8_t nameSpace = record[vo + 0x41];

                // Skip DOS-only namespace (2).
                if (nameSpace != 2 && nameLength > 0 && vo + 0x42 + nameLength * 2 <= recordLen) {
                    // Name is UTF-16LE.
                    std::u16string_view u16name(
                        reinterpret_cast<const char16_t*>(record.data() + vo + 0x42),
                        nameLength);

                    // Convert UTF-16 to UTF-8.
                    std::string utf8Name;
                    utf8Name.reserve(nameLength);
                    for (char16_t c : u16name) {
                        if (c < 0x80) {
                            utf8Name.push_back(static_cast<char>(c));
                        } else if (c < 0x800) {
                            utf8Name.push_back(static_cast<char>(0xC0 | (c >> 6)));
                            utf8Name.push_back(static_cast<char>(0x80 | (c & 0x3F)));
                        } else {
                            utf8Name.push_back(static_cast<char>(0xE0 | (c >> 12)));
                            utf8Name.push_back(static_cast<char>(0x80 | ((c >> 6) & 0x3F)));
                            utf8Name.push_back(static_cast<char>(0x80 | (c & 0x3F)));
                        }
                    }

                    MftNameEntry entry;
                    entry.ParentFrn = parentFrn;
                    entry.Name = std::move(utf8Name);
                    entry.FileAttributes = fileAttrs;
                    entry.RealSize = realSize;
                    entry.CreationUnix = creationUnix;
                    entry.LastWriteUnix = lastWriteUnix;
                    entry.LastAccessUnix = lastAccessUnix;
                    names.push_back(std::move(entry));
                }
            }
        } else if (attrType == 0x80) {
            // $DATA
            if (nameLen == 0) { // unnamed stream
                if (residentFlag == 0) {
                    // Non-resident: real size at +0x30 from attribute start.
                    if (attrOff + 0x30 + 8 <= recordLen) {
                        dataRealSize = ReadLE(record.data() + attrOff + 0x30, 8);
                    }
                } else {
                    // Resident: real size at +0x10.
                    uint16_t valueOff = static_cast<uint16_t>(ReadLE(record.data() + attrOff + 20, 2));
                    size_t vo = attrOff + valueOff;
                    if (vo + 8 <= recordLen) {
                        dataRealSize = ReadLE(record.data() + vo, 8);
                    }
                }
            }
        }

        attrOff += attrLen;
    }

    // Apply $DATA real size override.
    for (auto& name : names) {
        if (dataRealSize != UINT64_MAX) {
            name.RealSize = dataRealSize;
        }
    }

    return names;
}

} // namespace swiftlist::indexer

File: Core/Indexer/Mft/MftParser.cs
using System.Text;

namespace SwiftList.Core.Indexer.Mft;

/// <summary>Low-level NTFS $MFT structure parsing: fixup, $DATA run lists, and attribute walking.</summary>
internal static class MftParser
{
    /// <summary>Applies the update-sequence-array fixup to a raw FILE record in place.</summary>
    internal static void ApplyFixup(byte[] buf, uint bytesPerSector, int recOff, int recLen)
    {
        int usaOff = BitConverter.ToUInt16(buf, recOff + 4);
        int usaCount = BitConverter.ToUInt16(buf, recOff + 6);
        for (var i = 1; i < usaCount; i++)
        {
            var secEnd = recOff + i * (int)bytesPerSector - 2;
            var usaEntry = recOff + usaOff + i * 2;
            if (secEnd + 2 > recOff + recLen || usaEntry + 2 > buf.Length)
                break;
            buf[secEnd] = buf[usaEntry];
            buf[secEnd + 1] = buf[usaEntry + 1];
        }
    }

    /// <summary>Parses the $MFT's own non-resident $DATA run list into (lcn, clusterCount) extents.</summary>
    internal static List<(long lcn, long clusters)> ParseDataRuns(byte[] rec)
    {
        var extents = new List<(long, long)>();
        ParseDataRunsInto(rec, extents);
        return extents;
    }

    internal static void ParseDataRunsInto(byte[] rec, List<(long lcn, long clusters)> extents)
    {
        int a = BitConverter.ToUInt16(rec, 0x14);
        while (a + 8 <= rec.Length)
        {
            var type = BitConverter.ToUInt32(rec, a);
            if (type == 0xFFFFFFFF)
                break;
            var len = BitConverter.ToUInt32(rec, a + 4);
            if (len < 16 || a + len > rec.Length)
                break;
            if (type == 0x80 && rec[a + 8] == 1) // $DATA, non-resident
            {
                int mpOff = BitConverter.ToUInt16(rec, a + 0x20);
                var p = a + mpOff;
                long lcn = 0;
                while (p < a + len && rec[p] != 0)
                {
                    var hdr = rec[p++];
                    var lenBytes = hdr & 0x0F;
                    var offBytes = (hdr >> 4) & 0x0F;
                    if (lenBytes == 0)
                        break;
                    var runLen = ReadLE(rec, p, lenBytes);
                    p += lenBytes;
                    if (offBytes == 0)
                        continue; // sparse hole (unexpected for $MFT)
                    var runOff = ReadSignedLE(rec, p, offBytes);
                    p += offBytes;
                    lcn += runOff;
                    extents.Add((lcn, runLen));
                }
            }
            a += (int)len;
        }
    }

    /// <summary>Parses resident and non-resident $ATTRIBUTE_LIST (0x20) entries to find record indexes of extension records holding <paramref name="targetAttrType"/>.</summary>
    internal static List<ulong> ParseAttributeListRecordIndexes(byte[] rec, uint targetAttrType, Func<long, byte[], int, bool>? readAt = null, uint bytesPerCluster = 0)
    {
        var records = new List<ulong>();
        int a = BitConverter.ToUInt16(rec, 0x14);
        while (a + 8 <= rec.Length)
        {
            var type = BitConverter.ToUInt32(rec, a);
            if (type == 0xFFFFFFFF)
                break;
            var len = BitConverter.ToUInt32(rec, a + 4);
            if (len < 16 || a + len > rec.Length)
                break;
            if (type == 0x20) // $ATTRIBUTE_LIST
            {
                var resident = rec[a + 8] == 0;
                if (resident)
                {
                    var vo = BitConverter.ToUInt16(rec, a + 0x14);
                    var vl = BitConverter.ToUInt32(rec, a + 0x10);
                    var p = a + vo;
                    var end = p + (int)vl;
                    if (end <= rec.Length)
                    {
                        ParseAttributeListEntries(rec.AsSpan(p, (int)vl), targetAttrType, records);
                    }
                }
                else if (readAt != null && bytesPerCluster > 0 && a + 0x38 <= rec.Length)
                {
                    var realSize = BitConverter.ToInt64(rec, a + 0x30);
                    if (realSize > 0 && realSize <= 16 * 1024 * 1024) // sanity cap 16MB for attribute list
                    {
                        var attrExtents = new List<(long lcn, long clusters)>();
                        ParseDataRunsFromAttribute(rec, a, attrExtents);
                        var attrBuf = new byte[realSize];
                        long readOffset = 0;
                        var success = true;
                        foreach (var (lcn, clusters) in attrExtents)
                        {
                            var bytesToRead = (int)Math.Min(clusters * bytesPerCluster, realSize - readOffset);
                            if (bytesToRead <= 0) break;
                            var alignedReadBytes = (int)(clusters * bytesPerCluster);
                            var tempBuf = new byte[alignedReadBytes];
                            if (!readAt(lcn * bytesPerCluster, tempBuf, alignedReadBytes))
                            {
                                success = false;
                                break;
                            }
                            Buffer.BlockCopy(tempBuf, 0, attrBuf, (int)readOffset, bytesToRead);
                            readOffset += bytesToRead;
                        }
                        if (success && readOffset > 0)
                        {
                            ParseAttributeListEntries(attrBuf.AsSpan(0, (int)readOffset), targetAttrType, records);
                        }
                    }
                }
            }
            a += (int)len;
        }
        return records;
    }

    internal static void ParseDataRunsFromAttribute(byte[] rec, int attrOffset, List<(long lcn, long clusters)> extents)
    {
        var len = BitConverter.ToUInt32(rec, attrOffset + 4);
        int mpOff = BitConverter.ToUInt16(rec, attrOffset + 0x20);
        var p = attrOffset + mpOff;
        long lcn = 0;
        while (p < attrOffset + len && p < rec.Length && rec[p] != 0)
        {
            var hdr = rec[p++];
            var lenBytes = hdr & 0x0F;
            var offBytes = (hdr >> 4) & 0x0F;
            if (lenBytes == 0 || p + lenBytes > rec.Length)
                break;
            var runLen = ReadLE(rec, p, lenBytes);
            p += lenBytes;
            if (offBytes == 0)
                continue;
            if (p + offBytes > rec.Length)
                break;
            var runOff = ReadSignedLE(rec, p, offBytes);
            p += offBytes;
            lcn += runOff;
            extents.Add((lcn, runLen));
        }
    }

    private static void ParseAttributeListEntries(ReadOnlySpan<byte> buffer, uint targetAttrType, List<ulong> records)
    {
        var p = 0;
        while (p + 0x18 <= buffer.Length)
        {
            var entryType = BitConverter.ToUInt32(buffer.Slice(p, 4));
            var entryLen = BitConverter.ToUInt16(buffer.Slice(p + 0x04, 2));
            if (entryLen < 0x18 || p + entryLen > buffer.Length)
                break;
            if (entryType == targetAttrType)
            {
                var mftRef = BitConverter.ToUInt64(buffer.Slice(p + 0x10, 8));
                var recIdx = mftRef & 0xFFFFFFFFFFFF;
                if (recIdx > 0 && !records.Contains(recIdx))
                    records.Add(recIdx);
            }
            p += entryLen;
        }
    }


    /// <summary>
    /// Walks a FILE record's attributes: collects every resident $FILE_NAME (excluding DOS-only 8.3
    /// short names) as (parentReference, name, realSize) into <paramref name="names"/>, and returns the
    /// $STANDARD_INFORMATION file attributes (hidden/system/etc) plus its Creation/LastWrite/LastAccess
    /// FILETIMEs (shared by every hard-linked name, unlike size which $FILE_NAME tracks per link).
    /// </summary>
    internal static uint CollectNames(byte[] buf, int recOff, int recLen, List<(UInt128 parent, string name, long size)> names,
        out long creationTimeUtc, out long lastWriteTimeUtc, out long lastAccessTimeUtc)
    {
        uint stdAttrs = 0;
        creationTimeUtc = 0;
        lastWriteTimeUtc = 0;
        lastAccessTimeUtc = 0;
        // $FILE_NAME's own "real size" (read below, per name) is $DATA's real size DUPLICATED into the
        // directory-entry-like $FILE_NAME record for fast listing -- NTFS only refreshes that copy on
        // rename/move/link, so it can sit stale (often 0) for a file that was written once and never
        // touched again. $DATA's own real-size field (read here from the same already-loaded record,
        // no extra I/O) is the one NTFS keeps authoritative, so it wins whenever present.
        long? dataSize = null;
        (UInt128 parent, string name, long size)? dosFallbackName = null;
        int a = BitConverter.ToUInt16(buf, recOff + 0x14);
        while (a + 8 <= recLen)
        {
            var type = BitConverter.ToUInt32(buf, recOff + a);
            if (type == 0xFFFFFFFF)
                break;
            var len = BitConverter.ToUInt32(buf, recOff + a + 4);
            if (len < 16 || a + len > recLen)
                break;
            var resident = buf[recOff + a + 8] == 0;
            if (type == 0x10 && resident) // $STANDARD_INFORMATION
            {
                var vo = BitConverter.ToUInt16(buf, recOff + a + 0x14);
                if (a + vo + 0x24 <= recLen)
                {
                    creationTimeUtc = BitConverter.ToInt64(buf, recOff + a + vo + 0x00);
                    lastWriteTimeUtc = BitConverter.ToInt64(buf, recOff + a + vo + 0x08);
                    lastAccessTimeUtc = BitConverter.ToInt64(buf, recOff + a + vo + 0x18);
                    stdAttrs = BitConverter.ToUInt32(buf, recOff + a + vo + 0x20);
                }
            }
            else if (type == 0x30 && resident) // $FILE_NAME
            {
                var vo = BitConverter.ToUInt16(buf, recOff + a + 0x14);
                var vp = recOff + a + vo;
                if (vp + 0x42 <= recOff + recLen)
                {
                    var fnAttrs = BitConverter.ToUInt32(buf, vp + 0x38);
                    if ((fnAttrs & (uint)FileAttributes.Directory) != 0)
                        stdAttrs |= (uint)FileAttributes.Directory;

                    var ns = buf[vp + 0x41]; // 0=POSIX 1=Win32 2=DOS 3=Win32&DOS
                    UInt128 parent = (ulong)BitConverter.ToInt64(buf, vp);
                    var size = BitConverter.ToInt64(buf, vp + 0x30); // real (logical) size -- may be stale, see dataSize above
                    int nameLen = buf[vp + 0x40];
                    if (vp + 0x42 + nameLen * 2 <= recOff + recLen)
                    {
                        var parsedName = Encoding.Unicode.GetString(buf, vp + 0x42, nameLen * 2);
                        if (ns != 2)
                        {
                            names.Add((parent, parsedName, size));
                        }
                        else
                        {
                            dosFallbackName ??= (parent, parsedName, size);
                        }
                    }
                }
            }
            else if (type == 0x80 && buf[recOff + a + 0x09] == 0) // $DATA, unnamed stream only (skip alternate data streams)
            {
                if (resident)
                {
                    // Resident value length IS the file's content length -- no separate size field.
                    if (a + 0x14 <= recLen)
                        dataSize = BitConverter.ToUInt32(buf, recOff + a + 0x10);
                }
                else if (a + 0x38 <= recLen)
                {
                    dataSize = BitConverter.ToInt64(buf, recOff + a + 0x30); // real (logical) size, always current
                }
            }
            a += (int)len;
        }

        if (names.Count == 0 && dosFallbackName.HasValue)
        {
            names.Add(dosFallbackName.Value);
        }

        if (dataSize.HasValue)
        {
            for (var i = 0; i < names.Count; i++)
            {
                var (parent, name, _) = names[i];
                names[i] = (parent, name, dataSize.Value);
            }
        }

        return stdAttrs;
    }

    private static long ReadLE(byte[] b, int off, int n)
    {
        long v = 0;
        for (var i = 0; i < n; i++)
            v |= (long)b[off + i] << (8 * i);
        return v;
    }

    private static long ReadSignedLE(byte[] b, int off, int n)
    {
        var v = ReadLE(b, off, n);
        if (n < 8 && (b[off + n - 1] & 0x80) != 0)
            v |= -1L << (8 * n);
        return v;
    }
}
